spdk/ctest/nvme_discover_test.c

```diff
@@ -0,0 +1,175 @@
+//
+// (C) Copyright 2018-2019 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+#include <stdio.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <cmocka.h>
+
+#include "spdk/stdinc.h"
+#include "spdk/nvme.h"
+#include "spdk/env.h"
+
+#include "nvme_control.h"
+#include "nvme_control_common.h"
+
+/**
+ * ==============================
+ * nvme_control mock functions
+ * ==============================
+ */
+
+struct ret_t * __wrap_init_ret(void)
+{
+	printf("Mock init_ret() runnning...\n");
+	return mock_ptr_type(struct ret_t *);
+}
+
+void __wrap_cleanup(void){
+	printf("Mock cleanup() runnning...\n");
+}
+
+void __wrap_collect(struct ret_t *ret){
+	printf("Mock collect() runnning...\n");
+}
+
+/**
+ * ===================
+ * SPDK mock functions
+ * ===================
+ */
+
+int __wrap_spdk_nvme_probe(
+		const struct spdk_nvme_transport_id *trid,
+		void *cb_ctx,
+		spdk_nvme_probe_cb probe_cb,
+		spdk_nvme_attach_cb attach_cb,
+		spdk_nvme_remove_cb remove_cb)
+{
+	printf("Verifying expected parameters for spdk_nvme_probe...\n");
+
+	check_expected(probe_cb);
+	check_expected(attach_cb);
+
+	return mock_type(int);
+}
+
+static void test_nvme_discover_null_g_controllers(void **state)
+{
+	(void) state; /*unused*/
+	struct ret_t *rv, *ret;
+
+	/* ret_t values for mock init_ret */
+	printf("Initializing return value for init_ret...\n");
+	ret = test_malloc(sizeof(struct ret_t));
+	ret->rc = 0;
+	ret->ctrlrs = NULL;
+	ret->nss = NULL;
+	snprintf(ret->err, sizeof(ret->err), "none");
+
+	/* Define param checks and return values */
+	will_return(__wrap_init_ret, &ret);
+	will_return(__wrap_spdk_nvme_probe, 0);
+	expect_any(__wrap_spdk_nvme_probe, probe_cb);
+	expect_any(__wrap_spdk_nvme_probe, attach_cb);
+
+	rv = nvme_discover();
+	if (!rv) {
```

Comment:
shouldn't we be testing for not NULL here given we are testing a pointer to struct rather than int return value?